File: go-backend-interview-prep/04_interfaces_and_generics/07_generics_basics/main.py
# Python equivalent of Generics Basics (compare with main.go)
# Go generics (compile-time) ~ Python typing.TypeVar (type hints, runtime unchecked)

from typing import TypeVar, Callable

T = TypeVar("T")
U = TypeVar("U")


def map_fn(s: list[T], fn: Callable[[T], U]) -> list[U]:
    """Generic map -- same as Go Map[T, U]."""
    return [fn(v) for v in s]


def filter_fn(s: list[T], predicate: Callable[[T], bool]) -> list[T]:
    return [v for v in s if predicate(v)]


def contains(s: list[T], target: T) -> bool:
    """Python lists support 'in' natively, but this matches Go's Contains."""
    return target in s


class Stack:
    """Python has no generics enforcement at runtime.
    Type hints are for documentation and static checkers only."""

    def __init__(self):
        self.items: list = []

    def push(self, v) -> None:
        self.items.append(v)

    def pop(self) -> tuple:
        if not self.items:
            return None, False
        return self.items.pop(), True

    def __len__(self) -> int:
        return len(self.items)


def main():
    # --- Map ---
    print("--- Map ---")
    nums = [1, 2, 3, 4, 5]
    doubled = map_fn(nums, lambda n: n * 2)
    print("doubled:", doubled)

    strs = map_fn(nums, lambda n: f"#{n}")
    print("as strings:", strs)

    # --- Filter ---
    print("\n--- Filter ---")
    evens = filter_fn(nums, lambda n: n % 2 == 0)
    print("evens:", evens)

    words = ["Go", "Python", "Rust", "Go"]
    long = filter_fn(words, lambda s: len(s) > 2)
    print("long words:", long)

    # --- Contains ---
    print("\n--- Contains ---")
    print("has 3:", contains(nums, 3))
    print("has 9:", contains(nums, 9))
    print("has Go:", contains(words, "Go"))

    # --- Stack ---
    print("\n--- Stack[int] ---")
    int_stack = Stack()
    int_stack.push(10)
    int_stack.push(20)
    int_stack.push(30)
    print("len:", len(int_stack))
    v, ok = int_stack.pop()
    print(f"pop: {v} (ok={ok})")

    print("\n--- Stack[str] ---")
    str_stack = Stack()
    str_stack.push("hello")
    str_stack.push("world")
    s, ok = str_stack.pop()
    print(f"pop: {s!r} (ok={ok})")

    # Key difference:
    # Go generics are enforced at compile time. Type errors are caught before running.
    # Python type hints are NOT enforced at runtime. They are for tools like mypy.


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/04_interfaces_and_generics/07_generics_basics/main.go
package main

import "fmt"

// --- Generic function: Map ---
func Map[T any, U any](s []T, fn func(T) U) []U {
	result := make([]U, len(s))
	for i, v := range s {
		result[i] = fn(v)
	}
	return result
}

// --- Generic function: Filter ---
func Filter[T any](s []T, predicate func(T) bool) []T {
	var result []T
	for _, v := range s {
		if predicate(v) {
			result = append(result, v)
		}
	}
	return result
}

// --- Generic function: Contains (needs comparable) ---
func Contains[T comparable](s []T, target T) bool {
	for _, v := range s {
		if v == target {
			return true
		}
	}
	return false
}

// --- Generic type: Stack ---
type Stack[T any] struct {
	items []T
}

func (s *Stack[T]) Push(v T) {
	s.items = append(s.items, v)
}

func (s *Stack[T]) Pop() (T, bool) {
	if len(s.items) == 0 {
		var zero T
		return zero, false
	}
	last := s.items[len(s.items)-1]
	s.items = s.items[:len(s.items)-1]
	return last, true
}

func (s *Stack[T]) Len() int {
	return len(s.items)
}

func main() {
	// --- Generic Map ---
	fmt.Println("--- Map ---")
	nums := []int{1, 2, 3, 4, 5}
	doubled := Map(nums, func(n int) int { return n * 2 })
	fmt.Println("doubled:", doubled)

	strs := Map(nums, func(n int) string { return fmt.Sprintf("#%d", n) })
	fmt.Println("as strings:", strs)

	// --- Generic Filter ---
	fmt.Println("\n--- Filter ---")
	evens := Filter(nums, func(n int) bool { return n%2 == 0 })
	fmt.Println("evens:", evens)

	words := []string{"Go", "Python", "Rust", "Go"}
	long := Filter(words, func(s string) bool { return len(s) > 2 })
	fmt.Println("long words:", long)

	// --- Generic Contains ---
	fmt.Println("\n--- Contains ---")
	fmt.Println("has 3:", Contains(nums, 3))
	fmt.Println("has 9:", Contains(nums, 9))
	fmt.Println("has Go:", Contains(words, "Go"))

	// --- Generic Stack ---
	fmt.Println("\n--- Stack[int] ---")
	var intStack Stack[int]
	intStack.Push(10)
	intStack.Push(20)
	intStack.Push(30)
	fmt.Println("len:", intStack.Len())
	v, ok := intStack.Pop()
	fmt.Printf("pop: %d (ok=%t)\n", v, ok)

	fmt.Println("\n--- Stack[string] ---")
	var strStack Stack[string]
	strStack.Push("hello")
	strStack.Push("world")
	s, ok := strStack.Pop()
	fmt.Printf("pop: %q (ok=%t)\n", s, ok)
}
